projects/avodata/firstStage.py

```diff
@@ -0,0 +1,121 @@
+import numpy as np
+import math
+
+
+def factorial(n):
+    return int(math.gamma(n + 1))
+
+
+def n_choose_k(n, k):
+    return factorial(n) / (factorial(k) * factorial(n - k))
+
+
+def generate_synthetic_a(num_of_professions):
+    num_of_rows = int(n_choose_k(num_of_professions, 2))
+    A = np.zeros(shape=(num_of_rows, num_of_professions), dtype=np.float32)
+    row = 0
+
+    for i in range(num_of_professions):
+        for j in range(i + 1, num_of_professions):
+            A[row][i] = 1
+            A[row][j] = -1
+            row += 1
+
+    return A
+
+
+def generate_synthetic_c(m, i, a):
+    c = np.dot(a, m[i].transpose())
+    return c
+
+
+def solve(A, c):
+    row_in_m = np.linalg.lstsq(A, c, None)[0]
+    return row_in_m
+
+def build_m(num_of_professions, c_matrix, num_of_characteristics):
+    A = generate_synthetic_a(num_of_professions)
+    m = np.zeros((num_of_characteristics, num_of_professions), dtype=np.float32)
+    for i in range(num_of_characteristics):
+        c = c_matrix[i]
+        row_in_m = solve(A, c)
+        row_in_m = (row_in_m - np.min(row_in_m)) / np.ptp(row_in_m)
+        m[i] = row_in_m
+    return m
+
+
+
+def run_check (m):
+    num_of_professions = m.shape[1]
+    A = generate_synthetic_a(num_of_professions)
+
+    num_of_rows_in_m = m.shape[0]
+    new_m = np.zeros((num_of_rows_in_m, num_of_professions), dtype=np.float32)
+    c_matrix = np.zeros((num_of_rows_in_m, num_of_professions), dtype=np.float32)
+    for i in range(num_of_rows_in_m):
+        c = generate_synthetic_c(m, i, A)
+        c_matrix[i] = c
+        row_in_m = solve(A, c)
+        new_m[i] = row_in_m
+    Ax = np.dot(A, new_m.transpose())
+    errors = np.subtract(Ax, c_matrix.transpose())
+
+    return new_m, errors
+
+#
+# def find_c_and_a(m):
+#     # A function that receives a matrix m with size k*n, such that k is a characteristics of
```

Comment:
If we leave commented code - we also explain why it's commented and when it should be uncommented.
Alternatively, if you don't need the code, you can delete it. It's always stored in git's history.

Reply:
Cool. good to know :)